### PR TITLE
Permission transitions

### PR DESCRIPTION
**Pull request by tariqkurd-repo**
fix https://github.com/riscv/riscv-cheri/issues/173
fix https://github.com/riscv/riscv-cheri/issues/164

### CODE REVIEW

**Review by arichardson — CHANGES_REQUESTED** [review]
Thanks, I think this addresses almost all transitions, but I believe we have not defined what happens when you try to remove C from `Execute + Data & Cap RO`. Since that is not representable we also end up removing X. Similar for inifinite-cap, removing C also removes ASR.

**Comment by arichardson** [comment]
I also believe we can't remove R from capabilities with C and have to also drop C.

File: src/insns/acperm_32bit.adoc [hunk 1]
```diff
@@ -32,12 +32,35 @@ output capability has its tag set to 0 if `cs1` is sealed.
...
+. <<x_perm>> cannot be set without <<r_perm>> being set.
+.. Therefore removing <<r_perm>> also requires removing <<x_perm>>.
+. If removing <<c_perm>> with <<x_perm>> set cannot be represented, then also remove <<x_perm>>.
+.. This effects the _Execute + Data & Cap RO_ case from <<cap_perms_encoding32>>.
+
+NOTE: Failing to follow the rules above will clear *all* permissions.
+
+The XLENMAX=64 additional rule is:
+
+. Clearing <<x_perm>> also clears the <<m_bit>>
```
Comment (arichardson) [inline src/insns/acperm_32bit.adoc:61]: This is also true for RV32, we just can't "directly read the M bit"

**Review by arichardson — APPROVED** [review]
Thanks this looks good to me now, but I'd like @LawrenceEsswood to have a look as well before merging since the format is based on his proposal.

**Comment by LawrenceEsswood** [comment]
I don't believe the list is exhaustive for RV32. It seems to at least be missing what happens when you try make a cap with ASR but missing any other permission.

Also, while defining all unrepresentable permissions sets to transition to no permissions does address https://github.com/riscv/riscv-cheri/issues/173, I was hoping that we would take a less surprising approach and instead transition to the nearest useful cap. 

cap-description basically states what I was hoping. I think it can be described succinctly for RV32 as (applying these rules in order):

If you do not have ALL permissions, clear ASR
If you do not have R, clear C and X
If you have X and R, but not C, clear X

File: src/insns/acperm_32bit.adoc (hunk 1 above)
Comment (tariqkurd-repo) [reply src/insns/acperm_32bit.adoc]: fixed

**Comment by tariqkurd-repo** [comment]
please have another look @LawrenceEsswood

**Comment by andresag01** [comment]
@tariqkurd-repo: I replaced instances of XLENMAX with MXLEN and rebased on top of main because the branch was behind.

**Review by andresag01 — APPROVED** [review]

**Comment by LawrenceEsswood** [comment]
> please have another look @LawrenceEsswood

Too late(!), but yes, LGTM

**Pull request merged**